docs/proposals/08-shoot-additional-container-runtimes.md

```diff
@@ -0,0 +1,139 @@
+* It shall be possible to configure multiple low-level runtimes in Gardener clusters, on the Worker Group level.
+
+### Non-Goals
+
+* Exposing the configuration of the underlying runtimes to the user is beyond the scope of this proposal. The initial version will support a default common-sense configuration for each of the underlying runtimes. If required, future enhancements will make it possible to fine-tune the underlying runtimes.
+
+## Proposal
+
+Gardener today assumes that all supported operating systems have Docker pre-installed in the base image. Starting with Docker Engine 1.11, Docker itself was [refactored](https://www.docker.com/blog/docker-engine-1-11-runc/) and cleaned-up to be based on the [containerd](https://containerd.io/) library. The first phase would be to change the default Kubelet configuration as described [here](https://kubernetes.io/docs/setup/production-environment/container-runtimes/#containerd) so that Kubernetes would use containerd instead of the default Dockershim. This will be implemented for CoreOS, Ubuntu, and SuSE-JeOS. 
+
+Once Gardener clusters are migrated to use CRI, we will implement two Gardener extensions, providing gVisor and Kata Containers as options for Gardener landscapes. The `WorkerGroup` specification will be extended to allow specifying a list of additional required Runtimes for nodes in that group. For example:
+```yaml
+workers:
+- name: worker-b8jg5
+  machineType: m5.large
+  volumeType: gp2
+  volumeSize: 50Gi
+  autoScalerMin: 1
+  autoScalerMax: 2
+  maxSurge: 1
+  containerRuntimes:
+  - gvisor
+  - kata-containers
+  machineImage:
+    name: coreos
+    version: 2135.6.0
+```
+
+Each extension will need to address the following concern:
+
+1. Add the low-level runtime binaries to the worker nodes. Each extension should store the runtime binary in a way that is accessible to a node to download and install during startup.
+1. Hook the runtime binary into the containerd configuration file, so that the runtime becomes available to containerd.
+1. Apply a label to each node that allows identifying nodes where the runtime is available.
+1. Apply the relevant `RuntimeClass` to the Gardener cluster, to expose the functionality to users.
+1. Provide a `ValidatingWebhook` to catch invalid configurations. For example, Kata Containers on AWS requires a `machineType` of `i3.metal`, so any `Shoot` requests with a Kata Containers runtime and a different machine type on AWS should be rejected.
+
+Since each operating system distribution has different methods of installing software (apt on ubuntu, zypper on SuSE, Torcx on CoreOS/Flatcar), we will enhance the operating system extensions to add a standard `installSoftware` script which can abstract the mechanics of installing new software on a node. The new runtime extensions will use this new functionality to install the relevant runtimes to the worker nodes.
```

Comment:
See above and https://github.com/gardener/gardener/pull/1771#issuecomment-569601529 - we should only rely on container images if possible. We don't have a place nor automation to make arbitrary binaries accessible for all deployment types (public cloud, private cloud, on-prem). cc @ccwienk @vlerenc 

Can you please clarify how you envision `installSoftware` to look like?

Reply:
See above, combination of OS and container runtime and invention of a contract that is not sufficient.

Reply:
We agree. See https://github.com/gardener/gardener/pull/1735#discussion_r362012792 above.